ibis/util.py

```diff
@@ -355,3 +356,39 @@ def flatten_iterable(iterable):
             yield from flatten_iterable(item)
         else:
             yield item
+
+
+def warn_deprecated(name, ver='', instead='', ignoreframes=1):
+    """Warn about deprecated usage, including stacktrace, and
+    what to do instead."""
+    warnlines = []
+    msg = f'"{name}" is deprecated'
+    if ver:
+        msg += f' since v{ver}'
+
+    if instead:
+        msg += f'; use {instead}.'
+
+    warnlines.append(msg)
+
+    import traceback
+
+    warnlines.append('Deprecated call traceback (most recent last):')
+    for f in traceback.extract_stack(limit=6)[:-ignoreframes]:
+        warnlines.append(f'{f.filename} at line {f.lineno} in {f.name}')
+    warnings.warn('\n'.join(warnlines), DeprecationWarning)
+
+
+def deprecated(ver='', instead=''):
```

Comment:
Are these defaults meaningful? If I decorate without passing them, do I still get a useful error message?

I suspect that we should make these keyword-only arguments (and rename `ver` to `version`, setting it to `None` by default as well) + making `instead` a required keyword-only argument so that it's clear at the call site what the arguments mean. Requiring an `instead` also forces developers to document what to do when the warning is shown.

Reply:
`warn_deprecated` does compose a meaningful message if those arguments are omitted.  I can make `instead` be required and keyword-only.  I prefer a default of empty string for `version` instead of `None` (for typing and clarity reasons, since `''` isn't a reasonable version anyway) but am happy to make that change too.

Reply:
One issue with always requiring the keywords is that the maximum `instead` that can fit on a single 79-character `@deprecated` line is 32 chars (without keywords can be +16 chars).